packages/client/src/runtime/core/init/getEngineInstance.ts

```diff
@@ -0,0 +1,35 @@
+import { ClientEngineType, getClientEngineType } from '@prisma/internals'
+
+import { BinaryEngine, DataProxyEngine, EngineConfig, LibraryEngine } from '../engines'
+import { PrismaClientValidationError } from '../errors/PrismaClientValidationError'
+import { resolveDatasourceUrl } from './resolveDatasourceUrl'
+
+export function getEngineInstance(_engineConfig: EngineConfig) {
+  let url: string | undefined
+
+  try {
+    url = resolveDatasourceUrl({
+      inlineDatasources: _engineConfig.inlineDatasources,
+      overrideDatasources: _engineConfig.overrideDatasources,
+      env: { ..._engineConfig.env, ...process.env },
+      clientVersion: _engineConfig.clientVersion,
+    })
+  } catch {
+    // the error does not matter, but that means we don't have a valid url which
+    // means we can't use the DataProxyEngine and will default to LibraryEngine
+  }
+
+  const engineType = getClientEngineType(_engineConfig.generator!)
+
+  if (url?.startsWith('prisma://') || TARGET_ENGINE_TYPE === 'edge') {
```

Comment:
Notice the main change from this PR here: The switch to the Data Proxy engine is either dynamic and can happen with either Library or Binary engine builds; or the Data Proxy engine is completely locked when targeting edge runtimes.